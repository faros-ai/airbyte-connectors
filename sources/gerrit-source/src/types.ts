import {RunMode} from './streams/common';

export interface GerritConfig {
  readonly url: string;
  readonly authentication: GerritAuthentication;
  readonly reject_unauthorized?: boolean;
  readonly projects?: string[];
  readonly excluded_projects?: string[];
  readonly run_mode?: RunMode;
  readonly custom_streams?: string[];
  readonly fetch_reviews?: boolean;
  readonly fetch_comments?: boolean;
  readonly cutoff_days?: number;
  readonly bucket_id?: number;
  readonly bucket_total?: number;
  readonly page_size?: number;
  readonly timeout?: number;
  readonly concurrency_limit?: number;
  readonly start_date?: string;
  readonly end_date?: string;

  // Computed fields
  readonly startDate?: Date;
  readonly endDate?: Date;
}

export type GerritAuthentication =
  | HttpPasswordAuthentication
  | CookieAuthentication
  | GitCookieAuthentication;

export interface HttpPasswordAuthentication {
  readonly type: 'http_password';
  readonly username: string;
  readonly password: string;
}

export interface CookieAuthentication {
  readonly type: 'cookie';
  readonly cookie_value: string;
}

export interface GitCookieAuthentication {
  readonly type: 'git_cookie';
  readonly git_cookie_value: string;
}

// Gerrit API response types
export interface GerritProject {
  readonly id: string;
  readonly name?: string;
  readonly parent?: string;
  readonly description?: string;
  readonly state?: 'ACTIVE' | 'READ_ONLY' | 'HIDDEN';
  readonly branches?: {[key: string]: string};
  readonly labels?: {[key: string]: any};
  readonly web_links?: WebLink[];
}

export interface GerritChange {
  readonly id: string;
  readonly project: string;
  readonly branch: string;
  readonly topic?: string;
  readonly attention_set?: {[key: string]: AttentionSetInfo};
  readonly assignee?: AccountInfo;
  readonly hashtags?: string[];
  readonly change_id: string;
  readonly subject: string;
  readonly status: 'NEW' | 'MERGED' | 'ABANDONED';
  readonly created: string;
  readonly updated: string;
  readonly submitted?: string;
  readonly submitter?: AccountInfo;
  readonly starred?: boolean;
  readonly stars?: string[];
  readonly reviewed?: boolean;
  readonly submit_type?: string;
  readonly mergeable?: boolean;
  readonly submittable?: boolean;
  readonly insertions?: number;
  readonly deletions?: number;
  readonly total_comment_count?: number;
  readonly unresolved_comment_count?: number;
  readonly _number: number;
  readonly owner: AccountInfo;
  readonly actions?: {[key: string]: ActionInfo};
  readonly labels?: {[key: string]: LabelInfo};
  readonly permitted_labels?: {[key: string]: string[]};
  readonly removable_reviewers?: AccountInfo[];
  readonly reviewers?: {[key: string]: AccountInfo[]};
  readonly pending_reviewers?: AccountInfo[];
  readonly reviewer_updates?: ReviewerUpdateInfo[];
  readonly messages?: ChangeMessageInfo[];
  readonly current_revision?: string;
  readonly revisions?: {[key: string]: RevisionInfo};
  readonly tracking_ids?: TrackingIdInfo[];
  readonly _more_changes?: boolean;
  readonly problems?: ProblemInfo[];
  readonly is_private?: boolean;
  readonly work_in_progress?: boolean;
  readonly has_review_started?: boolean;
  readonly revert_of?: number;
  readonly submission_id?: string;
  readonly cherry_pick_of_change?: number;
  readonly cherry_pick_of_patch_set?: number;
  readonly contains_git_conflicts?: boolean;
}

export interface AccountInfo {
  readonly _account_id?: number;
  readonly name?: string;
  readonly display_name?: string;
  readonly email?: string;
  readonly secondary_emails?: string[];
  readonly username?: string;
  readonly avatars?: AvatarInfo[];
  readonly _more_accounts?: boolean;
  readonly status?: string;
  readonly inactive?: boolean;
  readonly tags?: string[];
}

export interface AttentionSetInfo {
  readonly account: AccountInfo;
  readonly last_update: string;
  readonly reason: string;
}

export interface ActionInfo {
  readonly method?: string;
  readonly label?: string;
  readonly title?: string;
  readonly enabled?: boolean;
}

export interface LabelInfo {
  readonly optional?: boolean;
  readonly approved?: AccountInfo;
  readonly rejected?: AccountInfo;
  readonly recommended?: AccountInfo;
  readonly disliked?: AccountInfo;
  readonly blocking?: boolean;
  readonly value?: number;
  readonly default_value?: number;
  readonly values?: {[key: string]: string};
  readonly all?: ApprovalInfo[];
}

export interface ApprovalInfo {
  readonly account?: AccountInfo;
  readonly value?: number;
  readonly permitted_voting_range?: VotingRangeInfo;
  readonly date?: string;
  readonly tag?: string;
  readonly post_submit?: boolean;
}

export interface VotingRangeInfo {
  readonly min: number;
  readonly max: number;
}

export interface ReviewerUpdateInfo {
  readonly updated: string;
  readonly updated_by: AccountInfo;
  readonly reviewer: AccountInfo;
  readonly state: 'REVIEWER' | 'CC' | 'REMOVED';
}

export interface ChangeMessageInfo {
  readonly id: string;
  readonly author?: AccountInfo;
  readonly real_author?: AccountInfo;
  readonly date: string;
  readonly message: string;
  readonly tag?: string;
  readonly _revision_number?: number;
}

export interface RevisionInfo {
  readonly kind: string;
  readonly _number: number;
  readonly created: string;
  readonly uploader: AccountInfo;
  readonly ref: string;
  readonly fetch?: {[key: string]: FetchInfo};
  readonly commit?: CommitInfo;
  readonly files?: {[key: string]: FileInfo};
  readonly actions?: {[key: string]: ActionInfo};
  readonly reviewed?: boolean;
  readonly commit_with_footers?: boolean;
  readonly push_certificate?: PushCertificateInfo;
  readonly description?: string;
}

export interface FetchInfo {
  readonly url: string;
  readonly ref: string;
  readonly commands?: {[key: string]: string};
}

export interface CommitInfo {
  readonly commit?: string;
  readonly parents: ParentCommitInfo[];
  readonly author: GitPersonInfo;
  readonly committer: GitPersonInfo;
  readonly subject: string;
  readonly message: string;
  readonly web_links?: WebLink[];
}

export interface ParentCommitInfo {
  readonly commit: string;
  readonly subject?: string;
}

export interface GitPersonInfo {
  readonly name: string;
  readonly email: string;
  readonly date: string;
  readonly tz: number;
}

export interface FileInfo {
  readonly status?: 'A' | 'D' | 'R' | 'C' | 'W';
  readonly binary?: boolean;
  readonly old_path?: string;
  readonly lines_inserted?: number;
  readonly lines_deleted?: number;
  readonly size_delta?: number;
  readonly size?: number;
}

export interface PushCertificateInfo {
  readonly certificate: string;
  readonly key: GpgKeyInfo;
}

export interface GpgKeyInfo {
  readonly status: string;
  readonly problems?: string[];
}

export interface TrackingIdInfo {
  readonly system: string;
  readonly id: string;
}

export interface ProblemInfo {
  readonly message: string;
  readonly status?: 'FIXED' | 'FIX_FAILED';
  readonly outcome?: string;
}

export interface AvatarInfo {
  readonly url: string;
  readonly height?: number;
  readonly width?: number;
}

export interface WebLink {
  readonly url: string;
  readonly name?: string;
  readonly image_url?: string;
  readonly target?: string;
}

export interface GerritGroup {
  readonly id: string;
  readonly name?: string;
  readonly url?: string;
  readonly options?: GroupOptionsInfo;
  readonly description?: string;
  readonly group_id?: number;
  readonly owner?: string;
  readonly owner_id?: string;
  readonly created_on?: string;
  readonly _more_groups?: boolean;
  readonly members?: AccountInfo[];
  readonly includes?: GroupInfo[];
}

export interface GroupInfo {
  readonly id: string;
  readonly name?: string;
  readonly url?: string;
  readonly options?: GroupOptionsInfo;
  readonly description?: string;
  readonly group_id?: number;
  readonly owner?: string;
  readonly owner_id?: string;
  readonly created_on?: string;
  readonly _more_groups?: boolean;
}

export interface GroupOptionsInfo {
  readonly visible_to_all?: boolean;
}
