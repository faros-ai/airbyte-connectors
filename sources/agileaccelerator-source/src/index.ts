import {Command} from 'commander';
import {
  AirbyteLogger,
  AirbyteSourceBase,
  AirbyteSourceRunner,
  AirbyteSpec,
  AirbyteStreamBase,
} from 'faros-airbyte-cdk';
import VError from 'verror';

import {
  Agileaccelerator,
  AgileacceleratorConfig,
} from './agileaccelerator/agileaccelerator';
import {Works} from './streams';

/** The main entry point. */
export function mainCommand(): Command {
  const logger = new AirbyteLogger();
  const source = new AgileacceleratorSource(logger);
  return new AirbyteSourceRunner(logger, source).mainCommand();
}

/** AgileAccelerator source implementation. */
export class AgileacceleratorSource extends AirbyteSourceBase<AgileacceleratorConfig> {
  async spec(): Promise<AirbyteSpec> {
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    return new AirbyteSpec(require('../resources/spec.json'));
  }
  async checkConnection(
    config: AgileacceleratorConfig
  ): Promise<[boolean, VError]> {
    try {
      const agileaccelerator = await Agileaccelerator.instance(
        config,
        this.logger
      );
      await agileaccelerator.checkConnection(config);
    } catch (err: any) {
      return [false, err];
    }
    return [true, undefined];
  }
  streams(config: AgileacceleratorConfig): AirbyteStreamBase[] {
    return [new Works(config, this.logger)];
  }
}
