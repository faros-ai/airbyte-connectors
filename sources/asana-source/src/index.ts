import {Command} from 'commander';
import {
  AirbyteSourceBase,
  AirbyteSourceRunner,
  AirbyteSpec,
  AirbyteStreamBase,
} from 'faros-airbyte-cdk';
import {AirbyteSourceLogger} from 'faros-airbyte-cdk';
import VError from 'verror';

import {Asana, AsanaConfig} from './asana';
import {
  Projects,
  ProjectTasks,
  Tags,
  Tasks,
  Users,
  Workspaces,
} from './streams';

/** The main entry point. */
export function mainCommand(): Command {
  const logger = new AirbyteSourceLogger();
  const source = new AsanaSource(logger);
  return new AirbyteSourceRunner(logger, source).mainCommand();
}

/** Asana source implementation. */
export class AsanaSource extends AirbyteSourceBase<AsanaConfig> {
  get type(): string {
    return 'asana';
  }

  async spec(): Promise<AirbyteSpec> {
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    return new AirbyteSpec(require('../resources/spec.json'));
  }
  async checkConnection(config: AsanaConfig): Promise<[boolean, VError]> {
    try {
      const asana = Asana.instance(config, this.logger);
      await asana.checkConnection();
    } catch (err: any) {
      return [false, err];
    }
    return [true, undefined];
  }
  streams(config: AsanaConfig): AirbyteStreamBase[] {
    return [
      new Projects(config, this.logger),
      new Tags(config, this.logger),
      new Tasks(config, this.logger),
      new Users(config, this.logger),
      new Workspaces(config, this.logger),
      new ProjectTasks(config, this.logger),
    ];
  }
}
