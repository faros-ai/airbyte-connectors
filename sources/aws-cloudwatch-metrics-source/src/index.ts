import {Command} from 'commander';
import {
  AirbyteSourceBase,
  AirbyteSourceLogger,
  AirbyteSourceRunner,
  AirbyteSpec,
  AirbyteStreamBase,
} from 'faros-airbyte-cdk';
import VError from 'verror';

import {CloudWatch, Config} from './cloudwatch';
import {Metrics} from './streams';

/** The main entry point. */
export function mainCommand(): Command {
  const logger = new AirbyteSourceLogger();
  const source = new CloudWatchMetricsSource(logger);
  return new AirbyteSourceRunner(logger, source).mainCommand();
}

/** CloudWatch Metrics source implementation. */
export class CloudWatchMetricsSource extends AirbyteSourceBase<Config> {
  get type(): string {
    return 'cloudwatch-metrics';
  }

  async spec(): Promise<AirbyteSpec> {
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    return new AirbyteSpec(require('../resources/spec.json'));
  }
  async checkConnection(config: Config): Promise<[boolean, VError]> {
    try {
      const cloudWatch = CloudWatch.instance(config);
      await cloudWatch.checkConnection();
    } catch (err: any) {
      return [false, err];
    }
    return [true, undefined];
  }
  streams(config: Config): AirbyteStreamBase[] {
    return [new Metrics(config, this.logger)];
  }
}
