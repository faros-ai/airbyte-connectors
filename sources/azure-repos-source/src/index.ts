import {Command} from 'commander';
import {
  AirbyteLogger,
  AirbyteSourceBase,
  AirbyteSourceRunner,
  AirbyteSpec,
  AirbyteStreamBase,
} from 'faros-airbyte-cdk';
import VError from 'verror';

import {AzureRepoConfig, AzureRepos} from './azure-repos';
import {PullRequests, Repositories, Users} from './streams';
import {Commits} from './streams/commits';

/** The main entry point. */
export function mainCommand(): Command {
  const logger = new AirbyteLogger();
  const source = new AzureRepoSource(logger);
  return new AirbyteSourceRunner(logger, source).mainCommand();
}

/** AzureRepo source implementation. */
export class AzureRepoSource extends AirbyteSourceBase<AzureRepoConfig> {
  async spec(): Promise<AirbyteSpec> {
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    return new AirbyteSpec(require('../resources/spec.json'));
  }
  async checkConnection(config: AzureRepoConfig): Promise<[boolean, VError]> {
    try {
      const azureRepos = await AzureRepos.make(config, this.logger);
      await azureRepos.checkConnection();
    } catch (err: any) {
      return [false, err];
    }
    return [true, undefined];
  }
  streams(config: AzureRepoConfig): AirbyteStreamBase[] {
    return [
      new Commits(config, this.logger),
      new PullRequests(config, this.logger),
      new Repositories(config, this.logger),
      new Users(config, this.logger),
    ];
  }
}
