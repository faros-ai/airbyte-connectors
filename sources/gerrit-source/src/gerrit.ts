import axios, {AxiosInstance, AxiosRequestConfig} from 'axios';
import {AirbyteLogger} from 'faros-airbyte-cdk';
import VError from 'verror';

import {
  AccountInfo,
  GerritChange,
  GerritConfig,
  GerritGroup,
  GerritProject,
} from './types';

export const DEFAULT_CUTOFF_DAYS = 90;
export const DEFAULT_RUN_MODE = 'Full';
export const DEFAULT_PAGE_SIZE = 100;
export const DEFAULT_TIMEOUT = 120000;

export interface ListProjectsOptions {
  limit?: number;
  start?: number;
  prefix?: string;
  regex?: string;
  branch?: string;
  description?: boolean;
  type?: 'code' | 'permissions' | 'all';
}

export interface ListChangesOptions {
  query?: string;
  limit?: number;
  start?: number;
  options?: string[];
}

export interface ListAccountsOptions {
  limit?: number;
  start?: number;
  query?: string;
}

export class Gerrit {
  private readonly client: AxiosInstance;
  private static instance_: Gerrit;

  constructor(
    readonly config: GerritConfig,
    readonly logger: AirbyteLogger
  ) {
    const auth = this.buildAuth(config.authentication);

    this.client = axios.create({
      baseURL: config.url.endsWith('/') ? config.url : `${config.url}/`,
      timeout: config.timeout ?? DEFAULT_TIMEOUT,
      httpsAgent:
        config.reject_unauthorized !== false
          ? undefined
          : {
              rejectUnauthorized: false,
            },
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json',
        ...auth.headers,
      },
      auth: auth.basicAuth,
    });

    // Add response interceptor to handle Gerrit's magic prefix
    this.client.interceptors.response.use((response) => {
      if (
        typeof response.data === 'string' &&
        response.data.startsWith(")]}'")
      ) {
        try {
          response.data = JSON.parse(response.data.slice(4));
        } catch (err) {
          this.logger.warn(`Failed to parse Gerrit response: ${err}`);
        }
      }
      return response;
    });

    // Add request interceptor for logging
    this.client.interceptors.request.use((config) => {
      this.logger.debug(
        `Gerrit API request: ${config.method?.toUpperCase()} ${config.url}`
      );
      return config;
    });
  }

  static async instance(
    config: GerritConfig,
    logger: AirbyteLogger
  ): Promise<Gerrit> {
    if (Gerrit.instance_) return Gerrit.instance_;

    Gerrit.instance_ = new Gerrit(config, logger);
    return Gerrit.instance_;
  }

  private buildAuth(auth: GerritConfig['authentication']): {
    headers?: Record<string, string>;
    basicAuth?: {username: string; password: string};
  } {
    switch (auth.type) {
      case 'http_password':
        return {
          basicAuth: {
            username: auth.username,
            password: auth.password,
          },
        };
      case 'cookie':
        return {
          headers: {
            Cookie: auth.cookie_value,
          },
        };
      case 'git_cookie': {
        // Parse git cookie format: hostname	FALSE	/	TRUE	0	o	git-username.domain.com=1//0XXXXXXXXXX
        const cookieValue = auth.git_cookie_value.trim();

        // Handle different cookie formats
        let cookie: string;
        if (cookieValue.includes('\t')) {
          // Tab-separated .gitcookies format: hostname	FALSE	/	TRUE	0	cookieName	cookieValue
          const parts = cookieValue.split('\t');
          if (parts.length >= 7) {
            const cookieName = parts[5];
            const cookieValuePart = parts[6];
            cookie = `${cookieName}=${cookieValuePart}`;
          } else {
            throw new VError(
              'Invalid .gitcookies format. Expected tab-separated format.'
            );
          }
        } else if (cookieValue.includes(',')) {
          // Comma-separated format: hostname,FALSE,/,TRUE,timestamp,cookieName,cookieValue
          const parts = cookieValue.split(',');
          if (parts.length >= 7) {
            const cookieName = parts[5];
            const cookieValuePart = parts[6];
            cookie = `${cookieName}=${cookieValuePart}`;
          } else {
            throw new VError(
              'Invalid comma-separated cookie format. Expected 7 fields.'
            );
          }
        } else if (cookieValue.includes('=')) {
          // Direct cookie format: cookieName=cookieValue
          cookie = cookieValue;
        } else {
          throw new VError(
            'Invalid git cookie format. Expected tab-separated, comma-separated, or direct cookie format.'
          );
        }

        return {
          headers: {
            Cookie: cookie,
          },
        };
      }
      default:
        throw new VError('Unsupported authentication type');
    }
  }

  private async request<T>(
    endpoint: string,
    config?: AxiosRequestConfig
  ): Promise<T> {
    try {
      // Use authenticated endpoint
      const url = endpoint.startsWith('/a/') ? endpoint : `/a${endpoint}`;
      const response = await this.client.get<T>(url, config);
      return response.data;
    } catch (err: any) {
      if (err.response) {
        throw new VError(
          {cause: err},
          `Gerrit API error: ${err.response.status} ${err.response.statusText}: ${JSON.stringify(err.response.data)}`
        );
      }
      throw new VError(
        {cause: err},
        `Gerrit API request failed: ${err.message}`
      );
    }
  }

  async getProjects(
    options: ListProjectsOptions = {}
  ): Promise<{[key: string]: GerritProject}> {
    const params: Record<string, any> = {};

    if (options.limit) params.n = options.limit;
    if (options.start) params.S = options.start;
    if (options.prefix) params.p = options.prefix;
    if (options.regex) params.r = options.regex;
    if (options.branch) params.b = options.branch;
    if (options.description) params.d = true;
    if (options.type) params.type = options.type;

    return this.request<{[key: string]: GerritProject}>('/projects/', {params});
  }

  async getChanges(options: ListChangesOptions = {}): Promise<GerritChange[]> {
    const params: Record<string, any> = {};

    if (options.query) params.q = options.query;
    if (options.limit) params.n = options.limit;
    if (options.start) params.S = options.start;
    if (options.options?.length) {
      // Add each option as a separate 'o' parameter
      options.options.forEach((option) => {
        if (!params.o) params.o = [];
        if (Array.isArray(params.o)) {
          params.o.push(option);
        } else {
          params.o = [params.o, option];
        }
      });
    }

    return this.request<GerritChange[]>('/changes/', {params});
  }

  async getAccounts(options: ListAccountsOptions = {}): Promise<AccountInfo[]> {
    const params: Record<string, any> = {};

    if (options.limit) params.n = options.limit;
    if (options.start) params.S = options.start;
    if (options.query) params.q = options.query;

    return this.request<AccountInfo[]>('/accounts/', {params});
  }

  async getGroups(
    options: {limit?: number; start?: number} = {}
  ): Promise<{[key: string]: GerritGroup}> {
    const params: Record<string, any> = {};

    if (options.limit) params.n = options.limit;
    if (options.start) params.S = options.start;

    return this.request<{[key: string]: GerritGroup}>('/groups/', {params});
  }

  async getProjectBranches(
    project: string,
    options: {limit?: number; start?: number} = {}
  ): Promise<any[]> {
    const params: Record<string, any> = {};

    if (options.limit) params.n = options.limit;
    if (options.start) params.S = options.start;

    const encodedProject = encodeURIComponent(project);
    return this.request<any[]>(`/projects/${encodedProject}/branches/`, {
      params,
    });
  }

  async getProjectTags(
    project: string,
    options: {limit?: number; start?: number} = {}
  ): Promise<any[]> {
    const params: Record<string, any> = {};

    if (options.limit) params.n = options.limit;
    if (options.start) params.S = options.start;

    const encodedProject = encodeURIComponent(project);
    return this.request<any[]>(`/projects/${encodedProject}/tags/`, {params});
  }
}
