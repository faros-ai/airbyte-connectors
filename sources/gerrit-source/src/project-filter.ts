import {AirbyteLogger} from 'faros-airbyte-cdk';

import {GerritConfig} from './types';

export class ProjectFilter {
  private static instance_: ProjectFilter;

  constructor(
    private readonly config: GerritConfig,
    private readonly logger: AirbyteLogger
  ) {}

  static instance(config: GerritConfig, logger: AirbyteLogger): ProjectFilter {
    if (ProjectFilter.instance_) return ProjectFilter.instance_;

    ProjectFilter.instance_ = new ProjectFilter(config, logger);
    return ProjectFilter.instance_;
  }

  shouldIncludeProject(projectName: string): boolean {
    // If specific projects are listed, only include those
    if (this.config.projects?.length) {
      return this.config.projects.includes(projectName);
    }

    // If project is in excluded list, skip it
    if (this.config.excluded_projects?.includes(projectName)) {
      return false;
    }

    // Include all other projects
    return true;
  }

  getProjectsToFetch(): string[] | undefined {
    return this.config.projects;
  }

  applyBucketing(projects: string[]): string[] {
    const bucketId = this.config.bucket_id ?? 1;
    const bucketTotal = this.config.bucket_total ?? 1;

    if (bucketTotal <= 1) {
      return projects;
    }

    this.logger.info(
      `Applying bucketing: bucket ${bucketId} of ${bucketTotal} total buckets`
    );

    return projects.filter((project, index) => {
      const bucket = (index % bucketTotal) + 1;
      return bucket === bucketId;
    });
  }
}
