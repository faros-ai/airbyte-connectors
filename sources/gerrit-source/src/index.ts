import {Command} from 'commander';
import {
  AirbyteConfiguredCatalog,
  AirbyteSourceBase,
  AirbyteSourceLogger,
  AirbyteSourceRunner,
  AirbyteSpec,
  AirbyteState,
  AirbyteStreamBase,
} from 'faros-airbyte-cdk';
import {calculateDateRange} from 'faros-airbyte-common/common';
import {VError} from 'verror';

import {GerritClient} from './gerrit';
import {FarosChanges} from './streams/faros_changes';
import {FarosProjects} from './streams/faros_projects';
import {FarosUsers} from './streams/faros_users';
import {GerritConfig} from './types';

export function mainCommand(): Command {
  const logger = new AirbyteSourceLogger();
  const source = new GerritSource(logger);
  return new AirbyteSourceRunner(logger, source).mainCommand();
}

export class GerritSource extends AirbyteSourceBase<GerritConfig> {
  get type(): string {
    return 'gerrit';
  }

  async spec(): Promise<AirbyteSpec> {
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    return new AirbyteSpec(require('../resources/spec.json'));
  }

  async checkConnection(config: GerritConfig): Promise<[boolean, any]> {
    try {
      const client = new GerritClient(config, this.logger);
      await client.checkConnection();
    } catch (err: any) {
      return [false, err];
    }
    return [true, undefined];
  }

  streams(config: GerritConfig): AirbyteStreamBase[] {
    return [
      new FarosProjects(config, this.logger),
      new FarosChanges(config, this.logger),
      new FarosUsers(config, this.logger),
    ];
  }

  async onBeforeRead(
    config: GerritConfig,
    catalog: AirbyteConfiguredCatalog,
    state?: AirbyteState
  ): Promise<{
    config: GerritConfig;
    catalog: AirbyteConfiguredCatalog;
    state?: AirbyteState;
  }> {
    // Calculate date range for incremental sync
    const {startDate, endDate} = calculateDateRange({
      start_date: config.start_date,
      end_date: config.end_date,
      cutoff_days: config.cutoff_days,
      logger: this.logger.info.bind(this.logger),
    });

    // Filter projects if specified
    let projectsToSync: string[] | undefined;
    if (config.projects?.length || config.excluded_projects?.length) {
      const client = new GerritClient(config, this.logger);
      const allProjects = new Set<string>();
      
      // Fetch all visible projects
      for await (const projectBatch of client.listProjects()) {
        for (const project of projectBatch) {
          allProjects.add(project.name);
        }
      }

      // Apply inclusion/exclusion filters
      const includedProjects = config.projects?.length 
        ? new Set(config.projects)
        : allProjects;
      
      const excludedProjects = new Set(config.excluded_projects || []);
      
      projectsToSync = [...includedProjects].filter(
        project => !excludedProjects.has(project)
      );

      if (projectsToSync.length === 0) {
        throw new VError(
          'No projects remain after applying inclusion and exclusion filters'
        );
      }

      this.logger.info(`Will sync ${projectsToSync.length} projects`);
    }

    return {
      config: {
        ...config,
        startDate,
        endDate,
        projects: projectsToSync,
        excluded_projects: undefined,
      },
      catalog,
      state,
    };
  }
}