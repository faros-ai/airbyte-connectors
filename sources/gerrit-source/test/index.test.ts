import {AirbyteLogger, AirbyteLogLevel, SyncMode} from 'faros-airbyte-cdk';
import {sourceCheckTest, sourceReadTest} from 'faros-airbyte-testing-tools';
import fs from 'fs';
import {GerritSource} from '../src';
import {GerritClient} from '../src/gerrit';

describe('gerrit source', () => {
  const logger = new AirbyteLogger(AirbyteLogLevel.WARN);
  const source = new GerritSource(logger);

  const testConfig = JSON.parse(fs.readFileSync('test/test-config.json', 'utf8'));

  beforeEach(() => {
    jest.clearAllMocks();
  });

  test('spec', async () => {
    await expect(source.spec()).resolves.toMatchSnapshot();
  });

  test('check connection - success', async () => {
    const checkConnection = jest.spyOn(GerritClient.prototype, 'checkConnection');
    checkConnection.mockResolvedValue(undefined);

    await sourceCheckTest({
      source,
      configOrPath: testConfig,
    });

    expect(checkConnection).toHaveBeenCalledTimes(1);
  });

  test('check connection - failure', async () => {
    const checkConnection = jest.spyOn(GerritClient.prototype, 'checkConnection');
    checkConnection.mockRejectedValue(new Error('test error'));

    await sourceCheckTest({
      source,
      configOrPath: testConfig,
    });

    expect(checkConnection).toHaveBeenCalledTimes(1);
  });

  test('read records', async () => {
    const listProjects = jest.spyOn(GerritClient.prototype, 'listProjects');
    listProjects.mockImplementation(async function* () {
      yield [{id: 'my-project', name: 'my-project'}];
    });


    const listChanges = jest.spyOn(GerritClient.prototype, 'listChanges');
    listChanges.mockImplementation(async function* () {
      yield [
        {
          _number: 123,
          project: 'my-project',
          subject: 'My Change',
          owner: {_account_id: 1, name: 'Test User', email: 'test@gerrit.test', username: 'testuser'},
          id: 'my-project~123',
          branch: 'main',
          change_id: 'I123abc',
          status: 'NEW',
          created: '2023-01-01T00:00:00Z',
          updated: '2023-01-01T00:00:00Z',
          insertions: 10,
          deletions: 5,
        },
      ];
    });

    const catalog = {
      streams: [
        {stream: {name: 'faros_projects', json_schema: {}}, sync_mode: SyncMode.INCREMENTAL},
        {stream: {name: 'faros_changes', json_schema: {}}, sync_mode: SyncMode.INCREMENTAL},
      ],
    };

    await sourceReadTest({
      source,
      configOrPath: testConfig,
      catalogOrPath: catalog,
      checkRecordsData: (records) => {
        expect(records.length).toBe(2);
        expect(records).toEqual(
          expect.arrayContaining([
            expect.objectContaining({name: 'my-project'}),
            expect.objectContaining({subject: 'My Change'}),
          ])
        );
      },
    });

    expect(listProjects).toHaveBeenCalledTimes(1);
    expect(listChanges).toHaveBeenCalledTimes(1);
  });
});