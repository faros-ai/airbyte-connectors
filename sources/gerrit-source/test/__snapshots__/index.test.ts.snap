// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gerrit source check connection - failure 1`] = `
AirbyteConnectionStatusMessage {
  "connectionStatus": {
    "message": "test error",
    "status": "FAILED",
  },
  "type": "CONNECTION_STATUS",
}
`;

exports[`gerrit source check connection - success 1`] = `
AirbyteConnectionStatusMessage {
  "connectionStatus": {
    "status": "SUCCEEDED",
  },
  "type": "CONNECTION_STATUS",
}
`;

exports[`gerrit source spec 1`] = `
AirbyteSpec {
  "spec": {
    "connectionSpecification": {
      "$schema": "http://json-schema.org/draft-07/schema#",
      "additionalProperties": true,
      "properties": {
        "authentication": {
          "description": "Choose your Gerrit authentication method",
          "oneOf": [
            {
              "properties": {
                "password": {
                  "airbyte_secret": true,
                  "description": "HTTP password from Gerrit account settings",
                  "order": 2,
                  "title": "HTTP Password",
                  "type": "string",
                },
                "type": {
                  "const": "basic",
                  "order": 0,
                  "type": "string",
                },
                "username": {
                  "description": "Gerrit username",
                  "order": 1,
                  "title": "Username",
                  "type": "string",
                },
              },
              "required": [
                "type",
                "username",
                "password",
              ],
              "title": "HTTP Basic Authentication",
              "type": "object",
            },
            {
              "properties": {
                "password": {
                  "airbyte_secret": true,
                  "description": "HTTP password from Gerrit account settings",
                  "order": 2,
                  "title": "HTTP Password",
                  "type": "string",
                },
                "type": {
                  "const": "digest",
                  "order": 0,
                  "type": "string",
                },
                "username": {
                  "description": "Gerrit username",
                  "order": 1,
                  "title": "Username",
                  "type": "string",
                },
              },
              "required": [
                "type",
                "username",
                "password",
              ],
              "title": "HTTP Digest Authentication",
              "type": "object",
            },
          ],
          "order": 1,
          "title": "Authentication",
          "type": "object",
        },
        "concurrency_limit": {
          "default": 4,
          "description": "Maximum number of concurrent API requests",
          "order": 8,
          "title": "Concurrency limit",
          "type": "integer",
        },
        "cutoff_days": {
          "default": 90,
          "description": "Only fetch data updated after cutoff",
          "order": 5,
          "title": "Cutoff Days",
          "type": "integer",
        },
        "end_date": {
          "description": "The date at which to stop syncing data. If not specified, uses the current date.",
          "format": "date-time",
          "order": 10,
          "title": "End Date",
          "type": "string",
        },
        "excluded_projects": {
          "description": "List of Gerrit projects from which data won't be fetched. By default, no projects are excluded.",
          "examples": [
            "test/experimental",
            "archive/old-project",
          ],
          "items": {
            "type": "string",
          },
          "order": 4,
          "title": "Excluded Projects list",
          "type": "array",
        },
        "page_size": {
          "default": 100,
          "description": "Maximum number of items in a paginated response",
          "order": 6,
          "title": "Page Size",
          "type": "integer",
        },
        "projects": {
          "description": "List of Gerrit projects from which to fetch data. If none passed, all visible projects will be used.",
          "examples": [
            "openstack/nova",
            "chromium/src",
          ],
          "items": {
            "type": "string",
          },
          "order": 3,
          "title": "Projects list",
          "type": "array",
        },
        "reject_unauthorized": {
          "default": true,
          "description": "Enable certificate validation for the Gerrit server",
          "order": 2,
          "title": "Enable certificate validation",
          "type": "boolean",
        },
        "server_url": {
          "description": "The URL of the Gerrit server (e.g., https://gerrit.example.com)",
          "examples": [
            "https://gerrit-review.googlesource.com",
            "https://review.opendev.org",
          ],
          "order": 0,
          "title": "Gerrit Server URL",
          "type": "string",
        },
        "start_date": {
          "description": "The date from which to start syncing data. If not specified, uses cutoff_days to calculate the start date.",
          "format": "date-time",
          "order": 9,
          "title": "Start Date",
          "type": "string",
        },
        "timeout": {
          "default": 120000,
          "description": "Timeout in milliseconds for each request to the Gerrit API",
          "order": 7,
          "title": "Request Timeout",
          "type": "integer",
        },
      },
      "required": [
        "server_url",
        "authentication",
      ],
      "title": "Gerrit Spec",
      "type": "object",
    },
    "documentationUrl": "https://docs.faros.ai",
  },
  "type": "SPEC",
}
`;
